bootstrap/src/stream.ml

```diff
@@ -0,0 +1,576 @@
+    rev_take (0 (1 (2 (3 Nil)))) 0 = Nil
+    rev_take (0 (1 (2 (3 Nil)))) 1 = (0 Nil)
+    rev_take (0 (1 (2 (3 Nil)))) 2 = (1 (0 Nil))
+    rev_take (0 (1 (2 (3 Nil)))) 3 = (2 (1 (0 Nil)))
+    rev_take (0 (1 (2 (3 Nil)))) 4 = (3 (2 (1 (0 Nil))))
+    rev_take (0 (1 (2 (3 Nil)))) 5 = (3 (2 (1 (0 Nil))))
+    rev_take (0 (1 (2 (3 (4 Nil))))) 0 = Nil
+    rev_take (0 (1 (2 (3 (4 Nil))))) 1 = (0 Nil)
+    rev_take (0 (1 (2 (3 (4 Nil))))) 2 = (1 (0 Nil))
+    rev_take (0 (1 (2 (3 (4 Nil))))) 3 = (2 (1 (0 Nil)))
+    rev_take (0 (1 (2 (3 (4 Nil))))) 4 = (3 (2 (1 (0 Nil))))
+    rev_take (0 (1 (2 (3 (4 Nil))))) 5 = (4 (3 (2 (1 (0 Nil)))))
+    rev_take (0 (1 (2 (3 (4 Nil))))) 6 = (4 (3 (2 (1 (0 Nil))))) |}]
+
+let%expect_test "drop" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let rec test_drop_up_to i l n =
+    match i <= l + 1, l <= n with
+    | _, false -> ()
+    | false, _ -> test_drop_up_to 0 (Usize.succ l) n
+    | true, true -> begin
+      let t = init l ~f:(fun i -> i) in
+      let t' = drop t i in
+      printf "drop %a %a = %a\n" ppt t Usize.pp i ppt t';
+      test_drop_up_to (Usize.succ i) l n
+    end in
+  test_drop_up_to 0 0 5;
+  printf "@]";
+
+  [%expect{|
+    drop Nil 0 = Nil
+    drop Nil 1 = Nil
+    drop (0 Nil) 0 = (0 Nil)
+    drop (0 Nil) 1 = Nil
+    drop (0 Nil) 2 = Nil
+    drop (0 (1 Nil)) 0 = (0 (1 Nil))
+    drop (0 (1 Nil)) 1 = (1 Nil)
+    drop (0 (1 Nil)) 2 = Nil
+    drop (0 (1 Nil)) 3 = Nil
+    drop (0 (1 (2 Nil))) 0 = (0 (1 (2 Nil)))
+    drop (0 (1 (2 Nil))) 1 = (1 (2 Nil))
+    drop (0 (1 (2 Nil))) 2 = (2 Nil)
+    drop (0 (1 (2 Nil))) 3 = Nil
+    drop (0 (1 (2 Nil))) 4 = Nil
+    drop (0 (1 (2 (3 Nil)))) 0 = (0 (1 (2 (3 Nil))))
+    drop (0 (1 (2 (3 Nil)))) 1 = (1 (2 (3 Nil)))
+    drop (0 (1 (2 (3 Nil)))) 2 = (2 (3 Nil))
+    drop (0 (1 (2 (3 Nil)))) 3 = (3 Nil)
+    drop (0 (1 (2 (3 Nil)))) 4 = Nil
+    drop (0 (1 (2 (3 Nil)))) 5 = Nil
+    drop (0 (1 (2 (3 (4 Nil))))) 0 = (0 (1 (2 (3 (4 Nil)))))
+    drop (0 (1 (2 (3 (4 Nil))))) 1 = (1 (2 (3 (4 Nil))))
+    drop (0 (1 (2 (3 (4 Nil))))) 2 = (2 (3 (4 Nil)))
+    drop (0 (1 (2 (3 (4 Nil))))) 3 = (3 (4 Nil))
+    drop (0 (1 (2 (3 (4 Nil))))) 4 = (4 Nil)
+    drop (0 (1 (2 (3 (4 Nil))))) 5 = Nil
+    drop (0 (1 (2 (3 (4 Nil))))) 6 = Nil
+    |}]
+
+let%expect_test "rev" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let rec test_rev_up_to i n = begin
+    match i <= n with
+    | false -> ()
+    | true -> begin
+      let t = init i ~f:(fun i' -> i') in
+      let t' = rev t in
+      printf "rev %a = %a\n" ppt t ppt t';
+      test_rev_up_to (Usize.succ i) n
+    end
+  end in
+  test_rev_up_to 0 5;
+  printf "@]";
+
+  [%expect{|
+    rev Nil = Nil
+    rev (0 Nil) = (0 Nil)
+    rev (0 (1 Nil)) = (1 (0 Nil))
+    rev (0 (1 (2 Nil))) = (2 (1 (0 Nil)))
+    rev (0 (1 (2 (3 Nil)))) = (3 (2 (1 (0 Nil))))
+    rev (0 (1 (2 (3 (4 Nil))))) = (4 (3 (2 (1 (0 Nil))))) |}]
```

Comment:
The tests look great in general.  Note that depending on the test, there's typically no real benefit to testing beyond a size in the 0..3 range.  Reducing the input ranges reduces what we have to manually review on subsequent changes, so it's a good idea to keep the output as small as is safe.